## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#9d1f3317-2d56-4961-b055-b3fb53563a10) |
| mint-large-bucket.sh | :x: [more...](#56c09664-47b1-4e72-8e10-347df416330c) |
| mint-dist-xl.sh | :x: [more...](#6eb55576-4823-4684-9339-08f0d3e9ae69) |
| mint-gateway-azure.sh | :x: [more...](#c965f62b-7e48-4841-9c78-ddd554bbcb0b) |
| mint-zoned.sh | :x: [more...](#18780867-c604-4eb8-8e04-75c552e46967) |
| mint-gateway-nas.sh | :x: [more...](#ececd430-eb01-4bfb-8170-82a31fab4bb2) |
<a name="18780867-c604-4eb8-8e04-75c552e46967"></a>
### 9972-d66a890/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev3.minio.io:30792
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 253b7f1b2bb5:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 19 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 53 seconds
(4/15) Running aws-sdk-ruby tests ... done in 11 seconds
(5/15) Running awscli tests ... done in 3 minutes and 24 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 28 seconds
(8/15) Running minio-dotnet tests ... done in 2 minutes and 34 seconds
(9/15) Running minio-go tests ... FAILED in 2 minutes and 44 seconds
{
  "args": {
    "bucketName": "minio-go-test-zwoqw4w431klfi9t",
    "objectName": "l1blfnxxatqrx63wpcviphjegq6z12",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 4296,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 5 minutes and 40 seconds
(10/15) Running minio-js tests ... done in 54 seconds
(11/15) Running minio-py tests ... done in 6 minutes and 4 seconds
(12/15) Running s3cmd tests ... done in 22 seconds
(13/15) Running s3select tests ... done in 10 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="ececd430-eb01-4bfb-8170-82a31fab4bb2"></a>
### 9972-d66a890/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:32363
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f3b6565ac421:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 0 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 33 seconds
(8/15) Running minio-dotnet tests ... done in 32 seconds
(9/15) Running minio-go tests ... done in 34 seconds
(10/15) Running minio-java tests ... FAILED in 8 seconds
{
  "name": "minio-java",
  "function": "uploadObject()",
  "args": "[multi-part upload]",
  "duration": 280,
  "status": "FAIL",
  "error": "java.net.SocketException: Connection reset by peer (Write failed) >>> [java.base/java.net.SocketOutputStream.socketWrite0(Native Method), java.base/java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:110), java.base/java.net.SocketOutputStream.write(SocketOutputStream.java:150), okio.Okio$1.write(Okio.java:79), okio.AsyncTimeout$1.write(AsyncTimeout.java:180), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:179), okio.RealBufferedSink.write(RealBufferedSink.java:42), okhttp3.internal.http1.Http1ExchangeCodec$KnownLengthSink.write(Http1ExchangeCodec.java:324), okio.ForwardingSink.write(ForwardingSink.java:35), okhttp3.internal.connection.Exchange$RequestBodySink.write(Exchange.java:231), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:179), okio.RealBufferedSink.write(RealBufferedSink.java:117), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:76), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:69), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229), okhttp3.RealCall.execute(RealCall.java:81), io.minio.MinioClient.execute(MinioClient.java:1014), io.minio.MinioClient.uploadPart(MinioClient.java:7781), io.minio.MinioClient.putObject(MinioClient.java:4706), io.minio.MinioClient.uploadObject(MinioClient.java:4931), FunctionalTest.testUploadObject(FunctionalTest.java:706), FunctionalTest.uploadObject_test(FunctionalTest.java:730), FunctionalTest.runTests(FunctionalTest.java:3791), FunctionalTest.main(FunctionalTest.java:4056)]"
}
(10/15) Running minio-js tests ... done in 47 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 31 seconds
(12/15) Running s3cmd tests ... done in 27 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="6eb55576-4823-4684-9339-08f0d3e9ae69"></a>
### 9972-d66a890/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31171
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0270292d5d37:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 14 seconds
(2/15) Running aws-sdk-java tests ... FAILED in 3 seconds
Jul 04, 2020 3:54:28 PM com.amazonaws.util.Base64 warn
WARNING: JAXB is unavailable. Will fallback to SDK implementation which may be less performant.If you are using Java 9+, you will need to include javax.xml.bind:jaxb-api as a dependency.
Exception in thread "main" com.amazonaws.services.s3.model.AmazonS3Exception: Server not initialized, please try again. (Service: Amazon S3; Status Code: 503; Error Code: XMinioServerNotInitialized; Request ID: 161E97A8E3A42FA7; S3 Extended Request ID: 47404f2b-46e4-4cce-8511-bfc92dca7372), S3 Extended Request ID: 47404f2b-46e4-4cce-8511-bfc92dca7372
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1742)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(AmazonHttpClient.java:1371)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1347)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1127)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:784)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:752)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:726)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:686)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:668)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:532)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:512)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5052)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4998)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4992)
	at com.amazonaws.services.s3.AmazonS3Client.listObjects(AmazonS3Client.java:895)
	at com.amazonaws.services.s3.AmazonS3Client.listObjects(AmazonS3Client.java:863)
	at io.minio.awssdk.tests.FunctionalTests.teardown(Unknown Source)
	at io.minio.awssdk.tests.FunctionalTests.main(Unknown Source)
(2/15) Running aws-sdk-php tests ... FAILED in 4 seconds
{
  "name": "aws-sdk-php",
  "function": "listObjects ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5f00a635e9654",
    "Object": "obj1"
  },
  "duration": "902",
  "status": "FAIL",
  "error": "Error executing \"ListObjects\" on \"http://minio-dev4.minio.io:31171/aws-sdk-php-5f00a635e9654-non-existent?encoding-type=url\"; AWS HTTP error: Server error: `GET http://minio-dev4.minio.io:31171/aws-sdk-php-5f00a635e9654-non-existent?encoding-type=url` resulted in a `503 Service Unavailable` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, pl (truncated...)\n XMinioServerNotInitialized (server): Server not initialized, please try again. - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><BucketName>aws-sdk-php-5f00a635e9654-non-existent</BucketName><Resource>/aws-sdk-php-5f00a635e9654-non-existent</Resource><RequestId>161E97A9BA8C81B8</RequestId><HostId>47404f2b-46e4-4cce-8511-bfc92dca7372</HostId></Error>"
}
(2/15) Running aws-sdk-ruby tests ... done in 9 seconds
(3/15) Running awscli tests ... FAILED in 20 seconds
{
  "name": "awscli",
  "duration": 15323,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-17183/datafile-1-kBremove_bucket failed: s3://awscli-mint-test-bucket-17183 An error occurred (XMinioServerNotInitialized) when calling the DeleteBucket operation (reached max retries: 4): Server not initialized, please try again."
}
(3/15) Running healthcheck tests ... done in 0 seconds
(4/15) Running mc tests ... FAILED in 1 minutes and 13 seconds
{
  "name": "mc",
  "duration": "73303",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-22385 --quiet --no-color mb myminio/mc-test-bucket-29438  >>> \nmc: <ERROR> Unable to make bucket `myminio/mc-test-bucket-29438`. Server not initialized, please try again."
}
(4/15) Running minio-dotnet tests ... done in 20 seconds
(5/15) Running minio-go tests ... FAILED in 1 minutes and 36 seconds
{
  "args": {
    "bucketName": "minio-go-test-hdo5y9qbby0tbih0"
  },
  "duration": 93404,
  "error": "Server not initialized, please try again.",
  "function": "GetObject(bucketName, objectName)",
  "message": "MakeBucket failed",
  "name": "minio-go: testGetObjectReadAtFunctionalV2",
  "status": "FAIL"
}
(5/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "makeBucket(MakeBucketArgs args)",
  "duration": 314,
  "status": "FAIL",
  "error": "java.lang.IllegalArgumentException: unknown error code string 'XMinioServerNotInitialized' >>> [io.minio.Xml.unmarshal(Xml.java:68), io.minio.MinioClient.execute(MinioClient.java:1050), io.minio.MinioClient.makeBucket(MinioClient.java:3773), FunctionalTest.makeBucket_test1(FunctionalTest.java:435), FunctionalTest.runTests(FunctionalTest.java:3772), FunctionalTest.main(FunctionalTest.java:4056)]"
}
(5/15) Running minio-js tests ... FAILED in 15 seconds
{
  "name": "minio-js",
  "function": "removeBucket(bucketName, cb)",
  "args": "bucketName:minio-js-test-c9647abc-4694-4e95-a691-4bdb87a617f2-region",
  "duration": 17,
  "status": "FAIL",
  "error": "S3Error: Server not initialized, please try again. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(5/15) Running minio-py tests ... FAILED in 11 minutes and 32 seconds
{
  "name": "minio-py:test_list_objects",
  "function": "list_objects(bucket_name, prefix, recursive)",
  "args": {
    "bucket_name": "minio-py-test-6e6e02ee-8098-483e-a3ac-181c4c290941",
    "object_name": "c9d5d092-6fbc-442c-a197-0f91af4363b3",
    "recursive": true
  },
  "duration": 606271,
  "message": "HTTPConnectionPool(host='minio-dev4.minio.io', port=31171): Max retries exceeded with url: /minio-py-test-6e6e02ee-8098-483e-a3ac-181c4c290941/?prefix= (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1035, in test_list_objects\n    for obj in objects:\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1036, in list_objects\n    headers=headers)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2000, in _url_open\n    preload_content=preload_content)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev4.minio.io', port=31171): Max retries exceeded with url: /minio-py-test-6e6e02ee-8098-483e-a3ac-181c4c290941/?prefix= (Caused by ResponseError('too many 503 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2085, in main\n    test_list_objects(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1042, in test_list_objects\n    raise Exception(err)\nException: HTTPConnectionPool(host='minio-dev4.minio.io', port=31171): Max retries exceeded with url: /minio-py-test-6e6e02ee-8098-483e-a3ac-181c4c290941/?prefix= (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(5/15) Running s3cmd tests ... FAILED in 52 seconds
{
  "name": "s3cmd",
  "duration": "49406",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "WARNING: Retrying failed request: /s3cmd-test-object-6812 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 3 sec...\nWARNING: Retrying failed request: /s3cmd-test-object-6812 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 6 sec...\nWARNING: Retrying failed request: /s3cmd-test-object-6812 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 9 sec...\nWARNING: Retrying failed request: /s3cmd-test-object-6812 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 12 sec...\nWARNING: Retrying failed request: /s3cmd-test-object-6812 (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 15 sec...\nERROR: S3 error: 503 (XMinioServerNotInitialized): Server not initialized, please try again."
}
(5/15) Running s3select tests ... FAILED in 10 minutes and 0 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "bucket_name": "s3select-test-d31514a7-e1c7-4f9b-b258-cf4c2c437d9e"
  },
  "duration": 600384,
  "message": "HTTPConnectionPool(host='minio-dev4.minio.io', port=31171): Max retries exceeded with url: /s3select-test-d31514a7-e1c7-4f9b-b258-cf4c2c437d9e/ (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 45, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 82, in test_csv_input_custom_quote_char\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 371, in make_bucket\n    headers=headers)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev4.minio.io', port=31171): Max retries exceeded with url: /s3select-test-d31514a7-e1c7-4f9b-b258-cf4c2c437d9e/ (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(5/15) Running security tests ... done in 0 seconds

Executed 5 out of 15 tests successfully.
```
<a name="56c09664-47b1-4e72-8e10-347df416330c"></a>
### 9972-d66a890/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32344
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f5d679a7bdc2:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 15 seconds
(2/15) Running aws-sdk-java tests ... done in 3 seconds
(3/15) Running aws-sdk-php tests ... done in 48 seconds
(4/15) Running aws-sdk-ruby tests ... done in 13 seconds
(5/15) Running awscli tests ... done in 2 minutes and 47 seconds
(6/15) Running healthcheck tests ... done in 4 seconds
(7/15) Running mc tests ... done in 1 minutes and 36 seconds
(8/15) Running minio-dotnet tests ... done in 2 minutes and 43 seconds
(9/15) Running minio-go tests ... FAILED in 2 minutes and 17 seconds
{
  "args": {
    "bucketName": "minio-go-test-1sx5b4e5ewdsg2qg",
    "objectName": "3eu56q63heg23mzokh0uemg6gx99s2",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 3855,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 5 minutes and 14 seconds
(10/15) Running minio-js tests ... done in 1 minutes and 0 seconds
(11/15) Running minio-py tests ... done in 4 minutes and 5 seconds
(12/15) Running s3cmd tests ... done in 28 seconds
(13/15) Running s3select tests ... done in 13 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="c965f62b-7e48-4841-9c78-ddd554bbcb0b"></a>
### 9972-d66a890/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31986
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 26349294783d:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 50 seconds
(4/15) Running aws-sdk-ruby tests ... done in 17 seconds
(5/15) Running awscli tests ... done in 2 minutes and 39 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 2 minutes and 58 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 34 seconds
(9/15) Running minio-go tests ... done in 4 minutes and 44 seconds
(10/15) Running minio-java tests ... FAILED in 5 minutes and 22 seconds
{
  "name": "minio-java",
  "function": "removeObjects(RemoveObjectsArgs args)",
  "duration": 1292,
  "status": "FAIL",
  "error": "org.simpleframework.xml.core.ValueRequiredException: Unable to satisfy @org.simpleframework.xml.Element(name=\"Key\", type=void.class, data=false, required=true) on field 'name' private java.lang.String io.minio.messages.DeletedObject.name for class io.minio.messages.DeletedObject at line 2 >>> [io.minio.Xml.unmarshal(Xml.java:68), io.minio.MinioClient.deleteObjects(MinioClient.java:7354), io.minio.MinioClient$1$1.populate(MinioClient.java:2945), io.minio.MinioClient$1$1.hasNext(MinioClient.java:2977), FunctionalTest.removeObjects(FunctionalTest.java:1395), FunctionalTest.removeObjects_test1(FunctionalTest.java:1534), FunctionalTest.runTests(FunctionalTest.java:3812), FunctionalTest.main(FunctionalTest.java:4056)]"
}
(10/15) Running minio-js tests ... done in 2 minutes and 32 seconds
(11/15) Running minio-py tests ... done in 9 minutes and 55 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 20 seconds
(13/15) Running s3select tests ... done in 1 minutes and 7 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="9d1f3317-2d56-4961-b055-b3fb53563a10"></a>
### 9972-d66a890/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32052
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b7c6bdeb9477:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 48 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 2 minutes and 19 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 43 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 3 seconds
(9/15) Running minio-go tests ... FAILED in 2 minutes and 6 seconds
{
  "args": {
    "bucketName": "minio-go-test-ktrzzug46zjpcpfm",
    "objectName": "haytqoj1zjvn51pbeyevfystr9mnc1",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 6161,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 2 minutes and 33 seconds
(10/15) Running minio-js tests ... done in 53 seconds
(11/15) Running minio-py tests ... done in 5 minutes and 19 seconds
(12/15) Running s3cmd tests ... done in 25 seconds
(13/15) Running s3select tests ... done in 8 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/data-crawler.go
+++ b/cmd/data-crawler.go
@@ -509,8 +509,9 @@ func (i *crawlItem) transformMetaDir() {
 
 // actionMeta contains information used to apply actions.
 type actionMeta struct {
-	oi      ObjectInfo
-	trustOI bool // Set true if oi can be trusted and has been read with quorum.
+	oi          ObjectInfo
+	trustOI     bool // Set true if oi can be trusted and has been read with quorum.
+	numVersions int  // The number of versions of this object
 }
 
 // applyActions will apply lifecycle checks on to a scanned item.
@@ -535,12 +536,13 @@ func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta action
 			VersionID:    meta.oi.VersionID,
 			DeleteMarker: meta.oi.DeleteMarker,
 			IsLatest:     meta.oi.IsLatest,
+			NumVersions:  meta.numVersions,
 		})
 	if i.debug {
 		logger.Info(color.Green("applyActions:")+" lifecycle: %q, Initial scan: %v", i.objectPath(), action)
 	}
 	switch action {
-	case lifecycle.DeleteAction:
+	case lifecycle.DeleteAction, lifecycle.DeleteVersionAction:
 	default:
 		// No action.
 		return size
@@ -586,14 +588,19 @@ func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta action
 		}
 		versionID = obj.VersionID
 		switch action {
-		case lifecycle.DeleteAction:
+		case lifecycle.DeleteAction, lifecycle.DeleteVersionAction:
 		default:
 			// No action.
 			return size
 		}
 	}
 
-	obj, err := o.DeleteObject(ctx, i.bucket, i.objectPath(), ObjectOptions{VersionID: versionID})
+	opts := ObjectOptions{}
+	if action == lifecycle.DeleteVersionAction {
+		opts.VersionID = versionID
+	}
+
+	obj, err := o.DeleteObject(ctx, i.bucket, i.objectPath(), opts)
 	if err != nil {
 		// Assume it is still there.
 		logger.LogIf(ctx, err)

--- a/cmd/data-crawler.go
+++ b/cmd/data-crawler.go
@@ -596,8 +596,11 @@ func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta action
 	}
 
 	opts := ObjectOptions{}
-	if action == lifecycle.DeleteVersionAction {
+	switch action {
+	case lifecycle.DeleteVersionAction:
 		opts.VersionID = versionID
+	case lifecycle.DeleteAction:
+		opts.Versioned = globalBucketVersioningSys.Enabled(i.bucket)
 	}
 
 	obj, err := o.DeleteObject(ctx, i.bucket, i.objectPath(), opts)

--- a/cmd/xl-storage.go
+++ b/cmd/xl-storage.go
@@ -408,7 +408,7 @@ func (s *xlStorage) CrawlAndGetDataUsage(ctx context.Context, cache dataUsageCac
 
 		var totalSize int64
 		for _, version := range fivs.Versions {
-			size := item.applyActions(ctx, objAPI, actionMeta{oi: version.ToObjectInfo(item.bucket, item.objectPath())})
+			size := item.applyActions(ctx, objAPI, actionMeta{numVersions: len(fivs.Versions), oi: version.ToObjectInfo(item.bucket, item.objectPath())})
 			if !version.Deleted {
 				totalSize += size
 			}

--- a/pkg/bucket/lifecycle/lifecycle.go
+++ b/pkg/bucket/lifecycle/lifecycle.go
@@ -209,8 +209,9 @@ func (lc Lifecycle) ComputeAction(obj ObjectOpts) Action {
 			return NoneAction
 		}
 
-		// All other expiration only applies to latest versions.
-		if obj.IsLatest {
+		// All other expiration only applies to latest versions
+		// (except if this is a delete marker)
+		if obj.IsLatest && !obj.DeleteMarker {
 			switch {
 			case !rule.Expiration.IsDateNull():
 				if time.Now().UTC().After(rule.Expiration.Date.Time) {

--- a/pkg/bucket/lifecycle/lifecycle.go
+++ b/pkg/bucket/lifecycle/lifecycle.go
@@ -40,6 +40,8 @@ const (
 	NoneAction Action = iota
 	// DeleteAction means the object needs to be removed after evaluting lifecycle rules
 	DeleteAction
+	// DeleteVersionAction deletes a particular version
+	DeleteVersionAction
 )
 
 // Lifecycle - Configuration for bucket lifecycle.
@@ -176,6 +178,7 @@ type ObjectOpts struct {
 	VersionID    string
 	IsLatest     bool
 	DeleteMarker bool
+	NumVersions  int
 }
 
 // ComputeAction returns the action to perform by evaluating all lifecycle rules
@@ -187,19 +190,19 @@ func (lc Lifecycle) ComputeAction(obj ObjectOpts) Action {
 	}
 
 	for _, rule := range lc.FilterActionableRules(obj) {
-		if obj.DeleteMarker && obj.IsLatest && bool(rule.Expiration.DeleteMarker) {
+		if obj.DeleteMarker && obj.NumVersions == 1 && bool(rule.Expiration.DeleteMarker) {
 			// Indicates whether MinIO will remove a delete marker with no noncurrent versions.
 			// Only latest marker is removed. If set to true, the delete marker will be expired;
 			// if set to false the policy takes no action. This cannot be specified with Days or
 			// Date in a Lifecycle Expiration Policy.
-			return DeleteAction
+			return DeleteVersionAction
 		}
 
 		if !rule.NoncurrentVersionExpiration.IsDaysNull() {
 			if obj.VersionID != "" && !obj.IsLatest {
 				// Non current versions should be deleted.
 				if time.Now().After(expectedExpiryTime(obj.ModTime, rule.NoncurrentVersionExpiration.NoncurrentDays)) {
-					return DeleteAction
+					return DeleteVersionAction
 				}
 				return NoneAction
 			}

